### PR TITLE
Download snapshots by most recent ledger index

### LINKED ISSUE #757: Download only the latest snapshot from the list of snapshot sources
When downloading a snapshot, Bee should only use the source that has the latest snapshot. This is already the case in Hornet.

https://github.com/gohornet/hornet/blob/ecdb18b0f8ea6a775807e1159e908d28a1ae0363/pkg/snapshot/download.go

### PR DESCRIPTION
**Pull request by Adam-Gleave**
# Description of change

Iterates through all provided snapshot sources, and uses streaming to download each `SnapshotHeader` to find the most recent ledger index. The snapshot with the most recent index will then be fully downloaded.

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

I have run the node with these changes and everything is working as expected. Appropriate errors are produced if the right conditions are met (for example, if there are no snapshot sources specified).

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Rebase and fix clippy please

File: bee-ledger/src/workers/snapshot/error.rs
Comment (samuel-rufi) [inline bee-ledger/src/workers/snapshot/error.rs]: we agreed on lowercase errors, right?

**Review by samuel-rufi — CHANGES_REQUESTED** [review]

File: bee-ledger/src/workers/snapshot/error.rs
Comment (Adam-Gleave) [reply bee-ledger/src/workers/snapshot/error.rs]: I believe so, but this whole crate has these sorts of errors 😬 Maybe another PR to sort that out?

File: bee-ledger/src/workers/snapshot/download.rs
Comment (pvdrz) [inline bee-ledger/src/workers/snapshot/download.rs]: `unwrap` wo comment :P

File: bee-ledger/src/workers/snapshot/download.rs
Comment (pvdrz) [inline bee-ledger/src/workers/snapshot/download.rs]: Could it be worth it to `Vec::with_capacity(SnapshotHeader::LENGTH)`?

**Review by pvdrz — DISMISSED** [review]

File: bee-ledger/src/workers/snapshot/download.rs
Comment (Adam-Gleave) [reply bee-ledger/src/workers/snapshot/download.rs]: Yes, I think so.

File: bee-ledger/src/workers/snapshot/download.rs
Comment (Adam-Gleave) [reply bee-ledger/src/workers/snapshot/download.rs]: 😬

**Review by pvdrz — DISMISSED** [review]

File: bee-ledger/src/workers/snapshot/error.rs
Comment (grtlr) [reply bee-ledger/src/workers/snapshot/error.rs]: Should we move this PR forward? I can fix the uppercase/lowercase errors later if they are blocking this.

**Comment by grtlr** [comment]
This PR would probably close issue #757.

File: bee-ledger/src/workers/snapshot/error.rs
Comment (grtlr) [reply bee-ledger/src/workers/snapshot/error.rs]: @thibault-martinez ?

Comment (thibault-martinez) [reply bee-ledger/src/workers/snapshot/error.rs]: You can fix it but it's not what's blocking this PR. There is a little issue with the way snapshots are selected, they should be selected by pairs.

File: bee-ledger/src/workers/snapshot/download.rs
Comment (pvdrz) [inline bee-ledger/src/workers/snapshot/download.rs]: can we avoid the closure?

**Review by pvdrz — DISMISSED** [review]

File: bee-ledger/CHANGELOG.md
Comment (thibault-martinez) [inline bee-ledger/CHANGELOG.md]: Update the date please

File: bee-ledger/src/workers/snapshot/error.rs
Comment (thibault-martinez) [inline bee-ledger/src/workers/snapshot/error.rs]: Can you lowercase all errors ?

File: bee-ledger/src/workers/snapshot/download.rs
Comment (grtlr) [inline bee-ledger/src/workers/snapshot/download.rs]: We could remove the `.unwrap` by using `if let ... =`.

File: bee-ledger/src/workers/snapshot/error.rs
Comment (grtlr) [reply bee-ledger/src/workers/snapshot/error.rs]: I'll also add dots. 😁

File: bee-ledger/src/workers/snapshot/import.rs
Comment (thibault-martinez) [inline bee-ledger/src/workers/snapshot/import.rs]: Is the map actually useful here ? We only evaluate the Some/None but never the content of the Some ?

File: bee-ledger/src/workers/snapshot/download.rs
Comment (Adam-Gleave) [inline bee-ledger/src/workers/snapshot/download.rs]: This isn't specific to this PR, but more of a general question: should our logging output follow the same convention as errors/panics etc, w.r.t. lowercase and not ending with punctuation? I think it makes sense to make it more consistent.

I believe `rustc` follows this convention for diagnostic messages, `env_log` and `tracing` both seem to do this in examples. However, the `log` crate examples (and `fern`) use uppercase characters, so it's not really consistent in the ecosystem either I guess 🤷 

Just a pedantic nitpick (and I was the one who wrote this line anyway!) but it just crossed my mind while looking at this.

File: bee-ledger/src/workers/snapshot/download.rs
Comment (Adam-Gleave) [inline bee-ledger/src/workers/snapshot/download.rs]: I wish these logs were one line :trollface:

File: bee-ledger/src/workers/snapshot/error.rs
Comment (Adam-Gleave) [reply bee-ledger/src/workers/snapshot/error.rs]: I think convention is also not to end an error message with punctuation (I hope you haven't added all those periods yet...😅)

File: bee-ledger/src/workers/snapshot/download.rs
Comment (thibault-martinez) [reply bee-ledger/src/workers/snapshot/download.rs]: Thing is, logging output actually always is a complete sentence, it always starts and ends with our string so it's pretty much always grammatically correct to start with an uppercase letter and end with a punctuation. Errors are often nested and included in already written panic message so it becomes weird to have uppercases and dots mixed in sentences.

File: bee-ledger/src/workers/snapshot/import.rs
Comment (thibault-martinez) [reply bee-ledger/src/workers/snapshot/import.rs]: `Path::exists` should actually be analysed to decide whether to import or not

Comment (thibault-martinez) [reply bee-ledger/src/workers/snapshot/import.rs]: `if let Some(true) = config.delta_path().map(Path::exists)` something like that ?

File: bee-ledger/src/workers/snapshot/download.rs
Comment (Adam-Gleave) [reply bee-ledger/src/workers/snapshot/download.rs]: Yep I can definitely see that, and I think that's the best reason for doing it this way. I don't have a strong opinion to be honest, but it just occured to me while I was reading through.

File: bee-ledger/src/workers/snapshot/download.rs
Comment (grtlr) [reply bee-ledger/src/workers/snapshot/download.rs]: That's the output of `cargo fmt`. I could add three lines of declaring shorter local variables before `warn!` :trollface:

File: bee-ledger/src/workers/snapshot/import.rs
Comment (grtlr) [reply bee-ledger/src/workers/snapshot/import.rs]: You are right, that's a bad one. 🙈

File: bee-ledger/src/workers/snapshot/download.rs
Comment (samuel-rufi) [inline bee-ledger/src/workers/snapshot/download.rs]: maybe change to `is_consistent`?

**Review by samuel-rufi — CHANGES_REQUESTED** [review]

File: bee-ledger/src/workers/snapshot/download.rs
Comment (grtlr) [inline bee-ledger/src/workers/snapshot/download.rs]: @thibault-martinez Does the sorting make sense? In Hornet, I think they only sort on the `ledger_index` of the full snapshot.

File: bee-ledger/src/workers/snapshot/download.rs
Comment (thibault-martinez) [inline bee-ledger/src/workers/snapshot/download.rs]: I wouldn't put this log as INFO, this could spam when having many sources and it's not very interesting to the node owner, maybe put it as debug.

File: bee-ledger/src/workers/snapshot/download.rs
Comment (thibault-martinez) [inline bee-ledger/src/workers/snapshot/download.rs]: I'd use `extend_from_slice` instead

File: bee-ledger/src/workers/snapshot/download.rs
Comment (thibault-martinez) [inline bee-ledger/src/workers/snapshot/download.rs]: Same for this log, debug

File: bee-ledger/src/workers/snapshot/download.rs
Comment (thibault-martinez) [inline bee-ledger/src/workers/snapshot/download.rs]: Log messages start with an uppercase and end with a punctuation :p

File: bee-ledger/src/workers/snapshot/download.rs
Comment (thibault-martinez) [inline bee-ledger/src/workers/snapshot/download.rs]:
```suggestion
                    "full snapshot SEP index does not match the delta snapshot ledger index ({} != {}): {}",
```

File: bee-ledger/src/workers/snapshot/download.rs
Comment (thibault-martinez) [inline bee-ledger/src/workers/snapshot/download.rs:188]: Just curious, what needs flattening here ?

File: bee-ledger/src/workers/snapshot/download.rs
Comment (thibault-martinez) [inline bee-ledger/src/workers/snapshot/download.rs:190]: Can we maybe use `retain` instead of filter/collect ? Less allocations

File: bee-ledger/src/workers/snapshot/download.rs
Comment (thibault-martinez) [reply bee-ledger/src/workers/snapshot/download.rs]: Let's do the same as hornet

File: bee-ledger/src/workers/snapshot/download.rs
Comment (grtlr) [reply bee-ledger/src/workers/snapshot/download.rs:188]: We get rid of all the sources that had an error when downloading the header. So essentially we go from `Iterator<Item=Result<SourceInformation>>` to `Iterator<Item=SourceInformation>`.

File: bee-ledger/src/workers/snapshot/download.rs
Comment (grtlr) [reply bee-ledger/src/workers/snapshot/download.rs:190]: Don't we have to collect anyways (unless we use something like `itertools`) because we need to sort afterwards?

File: bee-ledger/src/workers/snapshot/download.rs
Comment (thibault-martinez) [inline bee-ledger/src/workers/snapshot/download.rs:73]: I'm a bit confused, I thought you said in a previous comment that we should sort by ledger_index of full headers ?

File: bee-ledger/src/workers/snapshot/download.rs
Comment (thibault-martinez) [reply bee-ledger/src/workers/snapshot/download.rs:188]: Right, Result impls Iter, thanks

File: bee-ledger/src/workers/snapshot/download.rs
Comment (thibault-martinez) [reply bee-ledger/src/workers/snapshot/download.rs:190]: But can't we just modify `all_sources` with retain and then sort it ?

Comment (grtlr) [reply bee-ledger/src/workers/snapshot/download.rs:190]: Then we would also have to keep the `Result<..>`s and sort them as well, making the call to `.sort()` more complicated. Your call :trollface:

File: bee-ledger/src/workers/snapshot/download.rs
Comment (Adam-Gleave) [reply bee-ledger/src/workers/snapshot/download.rs]: Yeah that wasn't a criticism, it just makes me itch 😆

File: bee-ledger/src/workers/snapshot/download.rs
Comment (thibault-martinez) [reply bee-ledger/src/workers/snapshot/download.rs:190]: I see, forget about it then, it's just seeing all these copies that saddens me :(

File: bee-ledger/src/workers/snapshot/download.rs
Comment (grtlr) [reply bee-ledger/src/workers/snapshot/download.rs:73]: Sorry, I should have mentioned that: I checked with the code in Hornet again and they sort on the return of this method: https://github.com/gohornet/hornet/blob/ecdb18b0f8ea6a775807e1159e908d28a1ae0363/pkg/snapshot/snapshot_file.go#L270. Which should be the same as the method above (sans the check for `fullHeader`).

**Review by thibault-martinez — DISMISSED** [review]

File: bee-ledger/src/workers/snapshot/download.rs
Comment (thibault-martinez) [inline bee-ledger/src/workers/snapshot/download.rs]: I'm happy you did this, I wanted to ask for it but thought I would be way too pedantic

Comment (grtlr) [reply bee-ledger/src/workers/snapshot/download.rs]: 😊

**Pull request merged**
